spark3/src/test/java/org/apache/iceberg/spark/sql/TestAlterTable.java

```diff
@@ -225,4 +228,45 @@ public void testSetTableProperties() {
         UnsupportedOperationException.class,
         () -> sql("ALTER TABLE %s SET TBLPROPERTIES ('sort-order'='value')", tableName));
   }
+
+  @Test
+  public void testUpdateDataStoragePath() {
+    String objectStoragePath = "/folder/storage/path";
+    sql("ALTER TABLE %s SET TBLPROPERTIES ('%s'='true', '%s'='%s')",
+        tableName, TableProperties.OBJECT_STORE_ENABLED, TableProperties.OBJECT_STORE_PATH, objectStoragePath);
+
+    Table table = validationCatalog.loadTable(tableIdent);
+    LocationProvider locationProvider = table.locationProvider();
+    Assert.assertEquals("should use object storage location provider",
+        "org.apache.iceberg.LocationProviders$ObjectStoreLocationProvider",
+        locationProvider.getClass().getName());
+    Assert.assertTrue("should use table object storage path",
+        locationProvider.newDataLocation("file").contains(objectStoragePath));
+
+    String folderStoragePath = "/folder/storage/path";
+    sql("ALTER TABLE %s UNSET TBLPROPERTIES ('%s')",
+        tableName, TableProperties.OBJECT_STORE_PATH);
+    sql("ALTER TABLE %s SET TBLPROPERTIES ('%s'='%s')",
+        tableName, TableProperties.WRITE_NEW_DATA_LOCATION, folderStoragePath);
+
+    table.refresh();
+    locationProvider = table.locationProvider();
+    Assert.assertEquals("should use object storage location provider",
+        "org.apache.iceberg.LocationProviders$ObjectStoreLocationProvider",
+        locationProvider.getClass().getName());
+    Assert.assertTrue("should use table folder storage path",
```

Comment:
Nit: we might want to further clarify what we’re testing for in the assertion.

Something like `Should use table folder storage path after unsetting the object storage location path` or `should use table folder storage path if present when object storage path is not present”.

One could argue that these assertions could be subject to the same problems as comment rot if tests get changed, so I’ll defer to your judgement.

Also: Given that the names of the constants and their string representations are a little funky (particularly folder storage path / WRITE_NEW_DATA_LOCATION), it might make sense to refer to both at some point? Again, will leave that to your discretion but I think it might help clarify for readers. 🙂